---
title: "Getting Data into Bioconductor"
author: "Kasper D. Hansen"
---

```{r child="front.Rmd", echo=FALSE}
```

## Dependencies

This document has no dependencies.

```{r dependencies, warning=FALSE, message=FALSE}
```

## Overview

How do you get your data into R/Bioconductor?  The answer obviously depends on the file format of the data, but also what what you want to do with the data.  Generally speaking, you need access to the data file and then you need to put the data into a relevant **data container**.  Examples of data containers are `ExpressionSet` and `SummarizedExperiment`, but also classes such as `GRanges`.

Bioinformatics has jokingly been referred to as "The Science of Inventing New File Formats".  This joke exemplifies the myriad of different file formats in use.  Because we use many file formats and different types of data, it is hard to comprehensively cover all file formats and data types.

In general, a lot of useful solutions exists in domain / application specific packages.  As an example of this paradigm, the `r Biocpkg("affxparser")` package provides tools for parsing Affymetrix CEL files.  However, this package is a parsing library and returns the data in a less useful representation.  An end-user should instead use the `r Biocpkg("oligo")` package which uses `r Biocpkg("affxparser")` to read the data and then puts the data inside a useful data container; ready for downstream analysis.

## Application Area

### Microarray Data

Most microarray data is available to end users through a vendor specific file format such as CEL (Affymetrix) or IDAT  (Illumina).  These file formats can be read using vendor specific packages such as

- `r Biocpkg("affyio")`
- `r Biocpkg("affxparser")`
- `r Biocpkg("illuminaio")`

These packages are very low-level.  In practice, many analysis specific packages supports import of these files into useful data structures, and you are much better off using one of those packages.  For example

- `r Biocpkg("affy")` for Affymetrix Gene Expression data.
- `r Biocpkg("oligo")` for Affymetrix and Nimblegen expression and SNP array data.
- `r Biocpkg("lumi")` for Illumina arrays.
- `r Biocpkg("minfi")` for Illumina DNA methylation arrays (the 450k and 27k arrays).

### High-throughput sequencing

Raw (unmapped) reads are typically available in the FASTQ format.

The first step in most analyses is mapping the reads onto a genome.  For aligned reads, the BAM (SAM) format is now a clear standard.

However BAM (and SAM and FASTQ) files are quite big and still represents the data in a format which requires further processing before analysis.  However, this further processing vary by application area (ChIP, RNA, DNA etc).  Additionally, there are very few standard processed file formats; an example of such a standard format is BigWig.  As an example of the lack of standards, there is still no standard file format representing RNA-seq reads summarized at the gene or transcript level; different pipelines provide different sorts of file.  Luckily, these files are usually text files and can be read with standard tools for processing text files.ation from UCSC including UCSC tables can be accessed from the same package, for example by using the functions `getTable()` and `ucscTableQuery()`. 

There is also support for parsing GFF (Genome File Format) in `r Biocpkg("rtracklayer")`.

### VCF files

VCF (Variant Call Format) files represents genotype files, typically produced by running a genotyping pipeline on high-throughout sequencing data.  Use the functionality in `r Biocpkg("VariantAnnotation")` to access these files.

## Get data from databases of publicly available data

A number of data repositories have software packages dedicated to accessing the data inside of them:

- NCBI [GEO](http://www.ncbi.nlm.nih.gov/geo/) (Gene Expression Omnibus): the `r Biocpkg("GEOquery")` package.
- NCBI [SRA](http://www.ncbi.nlm.nih.gov/sra) (Short Read Archive): the `r Biocpkg("SRAdb")` package.
- EBI [ArrayExpress](https://www.ebi.ac.uk/arrayexpress/): the `r Biocpkg("ArrayExpress")` package.

## SessionInfo

\scriptsize

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

