# Syllabus / admin

- Get Syllabus into Github
	1. Add section on prerequisite - link to Roger's R course
    2. Link to reproducible research course
	3. Link to a data scientist toolsbox
	4. Add section on good skills to have and do 2. and 3.
	5. Something on laptop resources
- Write each week welcoming message
- Q10 finish
- Add wrong answers to assessment 4
- Add new due date to assessment 4
- Finish Youtube naming
- Add Youtube links to Github pages
- (Perhaps) add Coursera links to Github pages
- Add section to Syllabus explaining coursera hosted material vs. Github hosted material.
- Should all Rmds have an "edit on Github" link.
- Description using all_pkgs.R
- Go through all quiz'es and check that the right answer gives you 1pt.

# Remaining lectures

1. Getting your data into Bioconductor. DONE
2. limma DONE
3. oligo DONE
4. minfi
5. edgeR / DESeq
6. Rsamtools and a bit of ShortRead
7. Things to explore; ending the course

# Assessments

# Things left out

heatmaps
ggbio
VariantAnnotation
GenomicAlignments
Gviz / ggbio
GOstats
cummerbound

[CSHL](http://watson.nci.nih.gov/~sdavis/tutorials/IntroToR/)

## Massive scatterplots
- standard
- add loess line
- alpha blending
- smooth scatter
- heatmaps?


## Other list
ore

- Rsamtools / ShortRead
- GenomicAlignments
- GenomicFiles
- BiocParallel
- VariantAnnotation

Annotation

- biomaRt
- SNP pckages
- GWAS package

Visualization

- Gviz
- ggbio

Analysis

- limma
- DEseq / edgeR
- genefilter
- DNAcopy
- sva

Other

- Rsubread / Rbowtie
- RBGL / graph / Rgraphviz
- preprocessCore
- oligo
- GOstats
- EBImage


