---
title: "GenomeInfoDb"
author: "Kasper D. Hansen"
bibliography: genbioconductor.bib
---

```{r child="front.Rmd", echo=FALSE}
```

## Dependencies

This document has the following dependencies:

```{r dependencies, warning=FALSE, message=FALSE}
library(GenomeInfoDb)
library(GenomicRanges)
```

Use the following commands to install these packages in R.

```{r biocLite, eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite(c("GenomeInfoDb", "GenomicRanges"))
```

## Overview
  
The `r Biocpkg("GenomeInfoDb")` package addresses a seemingly small, but consistent problem: different online resources uses different naming conventions for chromosomes.  In more technical jargon, this package helps keeping your `seqinfo` and `seqlevels` harmonized.

## Other Resources

- The vignette from the [GenomeInfoDb webpage](http://bioconductor.org/packages/GenomeInfoDb).

## Changing style

It is an inconvenient truth that different online resources uses different naming convention for chromosomes.  This can even be different from organism to organism. For example, for the fruitfly (Drosophila Melanogaster) NCBI and Ensembl uses "2L" and UCSC uses "chr2L".  But NCBI and Ensembl differs on some contigs: NCBI uses "Un" and Ensembl used "U".

```{r GRanges}
gr <- GRanges(seqnames = "chr1", ranges = IRanges(start = 1:2, width = 2))
```

Let us remap
```{r seqStyle}
newStyle <- mapSeqlevels(seqlevels(gr), "NCBI")
gr <- renameSeqlevels(gr, newStyle)
```
This can in principle go wrong, if the original set of `seqlevels` are inconsistent (not a single style).

Hopefully, in Bioconductor 3.2 we will get support for `seqlengths` as well.

The `r Biocpkg("GenomeInfoDb")` also contains information for dropping / keeping various classes of chromosomes:

- `dropSeqlevels`
- `keepSeqlevels`
- `keepStandardChromosomes`


An easy way to drop chromosomes is by
```{r seqlevelsForce}
seqlevels(gr2, force = TRUE) <- "chr1"
gr2
```

FIXME: need to test and 

## SessionInfo

\scriptsize

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

## References
