---
title: "Getting Data into Bioconductor"
author: "Kasper D. Hansen"
bibliography: genbioconductor.bib
---

```{r child="front.Rmd", echo=FALSE}
```

## Dependencies

This document has the following dependencies:

```{r dependencies, warning=FALSE, message=FALSE}
library(GenomicRanges)
```

Use the following commands to install these packages in R.

```{r biocLite, eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite(c("GenomicRanges"))
```

## Overview

You should now have some idea of some useful data containers in Bioconductor (for example `ExpressionSet` and `SummarizedExperiment`) and you have learned some fundamentals of Bioconductor such as the `GRanges` class.

But how do you get your data into R/Bioconductor?  This depends a lot on the file format your data is in, and what you want to do with it.

Bioinformatics has jokingly been referred to as "The Science of Inventing New File Formats".  This exemplifies the myriad of different file formats in use across computational biology, and makes this task hard to cover quickly.

In this document we shall provide an overview.

## Other Resources

- The vignette from the [GenomicRanges webpage](http://bioconductor.org/packages/GenomicRanges).

## File formats

### Text files

Many different file formats are based on simple text files.  In case the text file is organized (essentially) as a matrix, you can read the file using `read.table()` from base R.  This function has a lot of arguments, and you are likely to have to change the default arguments for your purpose.  Important ones includes

- `sep`; what is the separate (typically either TAB, whitespace or `,`).
- `comment.char` and `quote`.  The default is to treat `#` and `'` in a special fashion, which can lead to parsing problems.  Especially `'` is encountered as 3' and 5'.
- `skip`; don't read the first `n` lines of the file.
- `header`; the first (read) line contains the column names.

Sometimes it 


`read.table`

`r CRANpkg("readr")`

`r CRANpkg("data.table")`

### Microarray Data

Most microarray data is available to end users through a vendor specific file format such as CEL (Affymetrix) or IDAT (Illumina).  These fileformats can be read using vendor specific packages such as

- `r Biopkg("affyio")`
- `r Biocpkg("affxparser")`
- `r Biocpkg("illuminaio")`

These packages are very low-level.  In practice, many analysis specific packages supports import of these files into useful data structures, and you are much better off using one of those packages.  For example

- `r Biocpkg("affy")` for Affymetrix Gene Expression data.
- `r Biocpkg("oligo")` for Affymetrix and Nimblegen Expression and SNP array data.
- `r Biopkg("lumi")` for Illumina arrays.
- `r Biocpkg("minfi")` for Illumina DNA methylation arrays (the 450k and 27k arrays).


### UCSC file formats

A number of file formats are invented by the UCSC genome browser.  For some applications, these are now standard file formats.  This includes

- BED files.
- BigBed files (large BED files, efficient access).
- Wig files.
- BigWig files (large Wig files, efficient access).

These files can be read using `r Biocpkg("rtracklayer")`, specifically the `import()` function.

Note: BED files looks like simple text files (and they are), but there are a number of conventions such as (1) they are zero-indexed and (2) all intervals are half-open.  These conventions are automatically mapped to the Bioconductor conventions (1) one-indexed and (2) all intervals are closed by `r Biocpkg("rtracklayer")`.

Other kinds of information from UCSC including UCSC tables can be accessed from the same package, for example by using the functions `getTable` and `ucscTableQuery`. 

There is also support for parsing GFF (Genome File Format) in `r Biocpkg("rtracklayer")`.

### FASTQ files

`r Biocpkg("ShortRead")`

### BAM files

BAM/SAM files represents aligned reads to a genome.  They are typically produced using a short read aligner such as Bowtie or MAQ or any of the many alternatives.

Use `r Biocpkg("Rsamtools")` to read these files.

### VCF files

VCF (Varient Call Format) files represents genotype files, typically produced by running a genotyping pipeline on high-throughout sequencing data.  Use the functionality in `r Biocpkg("VariantAnnotation")` to access these files.

### From databases of publicly available data

A number of data repositories have software packages dedicated to accesssing the data inside of them:

- [NCBI GEO](http://www.ncbi.nlm.nih.gov/geo/) (Gene Expression Omnibus): `r Biocpkg("GEOquery")`.
- [NCBI SRA](http://www.ncbi.nlm.nih.gov/sra) (Short Read Archive): `r Biocpkg("SRAdb")`.
- [EBI ArrayExpress](https://www.ebi.ac.uk/arrayexpress/): `r Biocpkg("ArrayExpress")`.






## SessionInfo

\scriptsize

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

## References
